protofsm/state_machine_test.go

```diff
@@ -0,0 +1,330 @@
+			}),
+		}, nil
+
+	// This state will proceed to the terminal state, but will emit all the
+	// possible daemon events.
+	case *daemonEvents:
+		// This send event can only succeed once the bool turns to
+		// true. After that, then we'll expect another event to take us
+		// to the final state.
+		sendEvent := &SendMsgEvent[dummyEvents]{
+			TargetPeer: *pub1,
+			SendWhen: fn.Some(func() bool {
+				return d.canSend.Load()
+			}),
+			PostSendEvent: fn.Some(dummyEvents(&goToFin{})),
+		}
+
+		// We'll also send out a normal send event that doesn't have
+		// any preconditions.
+		sendEvent2 := &SendMsgEvent[dummyEvents]{
+			TargetPeer: *pub2,
+		}
+
+		return &StateTransition[dummyEvents, *dummyEnv]{
+			// We'll state in this state until the send succeeds
+			// based on our predicate. Then it'll transition to the
+			// final state.
+			NextState: &dummyStateStart{
+				canSend: d.canSend,
+			},
+			NewEvents: fn.Some(EmittedEvent[dummyEvents]{
+				ExternalEvents: fn.Some(DaemonEventSet{
+					sendEvent, &DisableChannelEvent{},
+					&BroadcastTxn{}, sendEvent2,
+				}),
+			}),
+		}, nil
+	}
+
+	return nil, fmt.Errorf("unknown event: %T", event)
+}
+
+func (d *dummyStateStart) IsTerminal() bool {
+	return false
+}
+
+type dummyStateFin struct {
+}
+
+func (d *dummyStateFin) ProcessEvent(event dummyEvents, env *dummyEnv,
+) (*StateTransition[dummyEvents, *dummyEnv], error) {
+
+	return &StateTransition[dummyEvents, *dummyEnv]{
+		NextState: &dummyStateFin{},
+	}, nil
+}
+
+func (d *dummyStateFin) IsTerminal() bool {
+	return true
+}
+
+func assertState[Event any, Env Environment](t *testing.T,
+	m *StateMachine[Event, Env], expectedState State[Event, Env]) {
+
+	state, err := m.CurrentState()
+	require.NoError(t, err)
+	require.IsType(t, expectedState, state)
+}
+
+func assertStateTransitions[Event any, Env Environment](
+	t *testing.T, stateSub StateSubscriber[Event, Env],
+	expectedStates []State[Event, Env]) {
+
+	for _, expectedState := range expectedStates {
+		newState := <-stateSub.NewItemCreated.ChanOut()
+
+		require.IsType(t, expectedState, newState)
+	}
+}
+
+type dummyAdapters struct {
+	mock.Mock
+}
+
+func (d *dummyAdapters) SendMessages(pub btcec.PublicKey, msgs []lnwire.Message) error {
+	args := d.Called(pub, msgs)
+
+	return args.Error(0)
+}
+
+func (d *dummyAdapters) BroadcastTransaction(tx *wire.MsgTx, label string) error {
+	args := d.Called(tx, label)
+
+	return args.Error(0)
+}
+
+func (d *dummyAdapters) DisableChannel(op wire.OutPoint) error {
+	args := d.Called(op)
+
+	return args.Error(0)
+}
+
+// TestStateMachineTerminateCleanup tests that after the state machine
+// terminates, it properly cleans up the state machine.
+func TestStateMachineTerminateCleanup(t *testing.T) {
+	// First, we'll create our state machine given the env, and our
+	// starting state.
+	env := &dummyEnv{}
+	startingState := &dummyStateStart{}
+
+	adapters := &dummyAdapters{}
+
+	stateMachine := NewStateMachine[dummyEvents, *dummyEnv](
```

Comment:
eg yy, in the future this can be: 
```go
splicingStateMachine := NewStateMachine[splicing.Events, splicing.PeerEnv](...)
```

Then you count on this to execute the entire thing and react to all events. The one thing missing here is the layer one step above, that'll map any incoming wire messages or other daemon events into events in the domain of the state machine.

Reply:
@yyforyongyu here's what full init of some can look like: https://github.com/lightningnetwork/lnd/blob/43386d5643f961f948bf95513933c7d5a72fc74e/peer/brontide.go#L3098-L3109

Type aliases can be added to cut down on a bit of the line noise.